heron/statefulstorages/src/java/com/twitter/heron/statefulstorage/hdfs/HDFSStorage.java

```diff
@@ -0,0 +1,198 @@
+// Copyright 2017 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.statefulstorage.hdfs;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.twitter.heron.common.basics.SysUtils;
+import com.twitter.heron.proto.ckptmgr.CheckpointManager;
+import com.twitter.heron.spi.statefulstorage.Checkpoint;
+import com.twitter.heron.spi.statefulstorage.IStatefulStorage;
+
+/**
+ * Note: The hadoop cluster config should be provided through the classpath
+ */
+public class HDFSStorage implements IStatefulStorage {
+  private static final Logger LOG = Logger.getLogger(HDFSStorage.class.getName());
+
+  private static final String ROOT_PATH_KEY = "root.path";
+
+  private String checkpointRootPath;
+  private FileSystem fileSystem;
+
+  @Override
+  public void init(Map<String, Object> conf) {
+    LOG.info("Initialing... Config: " + conf.toString());
+    LOG.info("Class path: " + System.getProperty("java.class.path"));
+
+    checkpointRootPath = (String) conf.get(ROOT_PATH_KEY);
+
+    // Notice, we pass the config folder via classpath
+    // So hadoop will automatically search config files from classpath
+    Configuration hadoopConfig = new Configuration();
+
+    try {
+      fileSystem = FileSystem.get(hadoopConfig);
+      LOG.info("URI: " + fileSystem.getUri() + " ; Home Dir: " + fileSystem.getHomeDirectory());
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to get hadoop file system", e);
+    }
+  }
+
+  @Override
+  public void close() {
+    SysUtils.closeIgnoringExceptions(fileSystem);
+  }
+
+  @Override
+  public boolean store(Checkpoint checkpoint) {
+    Path path = new Path(getCheckpointPath(checkpoint));
+
+    // We need to ensure the existence of directories structure,
+    // since it is not guaranteed that FileSystem.create(..) always creates parents' dirs.
+    if (!ensureDirExists(getCheckpointDir(checkpoint))) {
```

Comment:
this should be part of initialization so we fail to start versus failing later upon first checkpoint.